core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala

```diff
@@ -124,11 +125,22 @@ private[spark] class ExternalSorter[K, V, C](
     if (shouldPartition) partitioner.get.getPartition(key) else 0
   }
 
+  private val metaInitialRecords = 256
+  private val kvChunkSize = conf.getInt("spark.shuffle.sort.kvChunkSize", 1 << 22) // 4 MB
+  private val useSerializedPairBuffer =
+    !ordering.isDefined && conf.getBoolean("spark.shuffle.sort.serializeMapOutputs", true) &&
+    ser.isInstanceOf[KryoSerializer] &&
```

Comment:
@JoshRosen definitely.  Filed https://issues.apache.org/jira/browse/SPARK-7311 for this.